numpy/core/src/multiarray/nditer_constr.c

```diff
@@ -2923,7 +2923,7 @@ npyiter_allocate_arrays(NpyIter *iter,
             /* If the data will be written to, set UPDATEIFCOPY */
             if (op_itflags[iop] & NPY_OP_ITFLAG_WRITE) {
                 Py_INCREF(op[iop]);
-                if (PyArray_SetUpdateIfCopyBase(temp, op[iop]) < 0) {
+                if (PyArray_SetWritebackIfCopyBase(temp, op[iop]) < 0) {
```

Comment:
Is this change and the next one, inside the `nditer` code, correct? I thought we were keeping `nditer` using `UPDATEIFCOPY` for now?

Reply:
My goal was for ``nditer`` to not print any warning after this pull request on CPython. The following logic allows that:
- all calls to ``PyArray_SetUpdateIfCopyBase`` issue a ``FutureWarning``,
- to maintain backward compatiblity, ``PyArray_ResolveWritebackIfCopy`` is called in ``array_dealloc`` if ``arr->base`` exists
  - if ``PyArray_ResolveWritebackIfCopy``actually does something, it will currently warn, but only on PyPy or if compiled with ``-DDEPRECATE_UPDATEIFCOPY" 

By using ``PyArray_SetWritebackIfCopyBase``, ``nditer`` will neither warn when creating its operands nor when resolving in ``array_dealloc``. When we deal with ``nditer`` in a future pull request, we can enable the warning in ``array_dealloc`` on all implementations.

Reply:
> all calls to `PyArray_SetUpdateIfCopyBase` issue a `FutureWarning`

This should be a `DeprecationWarning`. (Rule: `DeprecationWarning` is for when we're trying to remove an API, `FutureWarning` is for when we're going to keep it but change the semantics.) And I think this is where we should do the "if PyPy or `-DDEPRECATE_UPDATEIFCOPY`" check.

> `array_dealloc`

I think I'd prefer the `array_dealloc` logic be:

* If `WRITEBACKIFCOPY` is set, then issue a `RuntimeWarning` and that's it (you're supposed to call `ResolveUpdateIfCopy` or `ClearUpdateIfCopy` before the array is GCed).

* If `UPDATEIFCOPY` is set, then resolve it up like we always did. We don't need to issue a warning, because we already did that when `UPDATEIFCOPY` was set.

This then suggests that `nditer` should keep using `UPDATEIFCOPY` until we're ready to fix it properly, which seems logical, since it will in fact be broken until then, and if you're on PyPy or have `-DDEPRECATE_UPDATEIFCOPY` set then it's probably useful to see that your program is in fact using `UPDATEIFCOPY` (via nditer).